Source/DafnyCore/Substituter.cs

```diff
@@ -930,7 +979,73 @@ public class Substituter {
           }
         }
       }
+
       return attrs;
     }
+
+    /// <summary>
+    /// Substitution in a comprehension expression. 
+    /// Parameter 'forceSubstituteOfBoundVars' should be set to false if and only if
+    /// the expression is a binding guard, in which case a bound variable is introduced.
+    /// Such a variable must not be substituted. 
+    /// </summary>
+    protected Expression SubstituteComprehensionExpr(ComprehensionExpr expr, bool forceSubstituteOfBoundVars) {
+
+      Expression newExpr = null;
+
+      var e = (ComprehensionExpr)expr;
+      // For quantifiers and setComprehesion we want to make sure that we don't introduce name clashes with
+      // the enclosing scopes.
+
+      var q = e as QuantifierExpr;
+      if (q != null && q.SplitQuantifier != null) {
+        if (forceSubstituteOfBoundVars) {
+          return Substitute(q.SplitQuantifierExpression);
+        } else {
+          return SubstituteComprehensionExpr((ComprehensionExpr)q.SplitQuantifierExpression, false);
+        }
```

Comment:
As written, I think this code could be simplified to the following, because if `q.SplitQuantifierExpression` is a `ComprehensionExpr`, then the call to Substitute will immediately call `SubstituteComprehensionExpr(…, true)`.

```csharp
return SubstituteComprehensionExpr((ComprehensionExpr)q.SplitQuantifierExpression, forceSubstituteOfBoundVars);
```

However, I'm not sure this code is correct, because `q.SplitQuantifierExpression` is a conjunction of quantifiers, not a quantifier itself.  I would expect this code to raise an exception if it encounters a split quantifier — not just your new code, but also the old code.

This is because `Substitute(q.SplitQuantifierExpression);` was always wrong: it takes a quantifier and returns a conjunction or disjunction of quantifiers, so when applied to the `BindingGuard` of an `If` it replaces an `ExistsExpr` with a `BinaryExpr`, which confuses later parts of the code.  Here's a repro of this:

```
function method f(x: int): int { 10 - x * x }

function method BindingGuardTest(): int {
  var x: nat := 1;
  assert true by {
    if i :| 0 <= i < 10 && (f(i) == f(i+1) || f(i) == f(i+2)) {
      // print i;
    }
  }
  2
}
```

which outputs:

```
System.InvalidCastException: Unable to cast object of type 'Microsoft.Dafny.BinaryExpr' to type 'Microsoft.Dafny.ExistsExpr'.
   at Microsoft.Dafny.Translator.TrIfStmt(IfStmt stmt, BoogieStmtListBuilder builder, List`1 locals, ExpressionTranslator etran) 
```

For the bottom branch the fix would be to ignore the split expressions, I think.  But for the top branch if we want to preserve the splits then I think we have to apply the substitution on each of them.  @RustanLeino , thoughts?

Reply:
I am going to apply the simplification that you suggest. As for the issue with the split quantifiers, since it is a different issue, I will open a new issue to report it.

Reply:
I created a new issue #2748 to continue this discussion.